### ISSUE TITLE
cookiepatcher is changing .cookiecutterrc

### ISSUE THREAD

**Issue by dHannasch** [ionelmc/python-cookiepatcher#2]
https://gitlab.com/library-cookiecutters/cookiecutter-pylibrary/-/jobs/998305061
```
$ pip install git+https://github.com/cookiecutter/cookiecutter.git
Collecting git+https://github.com/cookiecutter/cookiecutter.git
  Cloning https://github.com/cookiecutter/cookiecutter.git to /tmp/pip-req-build-e64wqohw
  Running command git clone -q https://github.com/cookiecutter/cookiecutter.git /tmp/pip-req-build-e64wqohw
Collecting poyo>=0.5.0
  Downloading poyo-0.5.0-py2.py3-none-any.whl (10 kB)
Installing collected packages: poyo
Successfully installed poyo-0.5.0
$ git checkout $BRANCH_COOKIECUT_TO
Already on 'branch-to-run'
Your branch is up to date with 'origin/branch-to-run'.
$ cd ..
$ ls $repo_name/.cookiecutterrc
python-nameless/.cookiecutterrc
$ cookiepatcher ./ $repo_name --no-input
INFO: Loading config from 'python-nameless/.cookiecutterrc'
$ cd $repo_name
$ git diff
diff --git a/.cookiecutterrc b/.cookiecutterrc
index 0178cc5..d871c31 100644
--- a/.cookiecutterrc
+++ b/.cookiecutterrc
@@ -1,8 +1,8 @@
 # Generated by cookiepatcher, a small shim around cookiecutter (pip install cookiepatcher)
 
 default_context:
-    allow_tests_inside_package: yes
-    appveyor: no
+    allow_tests_inside_package: no
+    appveyor: yes
     c_extension_function: longest
     c_extension_module: _nameless
     c_extension_optional: no
@@ -12,7 +12,7 @@ default_context:
     codacy: no
     codacy_projectid: '[Get ID from https://app.codacy.com/app/library-cookiecutters/python-nameless/settings]'
     codeclimate: no
-    codecov: no
+    codecov: yes
     command_line_interface: click
     command_line_interface_bin_name: nameless
     coveralls: no
@@ -28,16 +28,16 @@ default_context:
     pre_commit: yes
     project_name: Nameless
     project_short_description: An example package. Generated with cookiecutter-pylibrary.
-    pypi_badge: no
+    pypi_badge: yes
     pypi_disable_upload: no
     release_date: today
     repo_hosting: other domain not listed
     repo_hosting_domain: gitlab.com
     repo_name: python-nameless
     repo_username: library-cookiecutters
-    requiresio: no
+    requiresio: yes
     scrutinizer: no
-    setup_py_uses_setuptools_scm: yes
+    setup_py_uses_setuptools_scm: no
     setup_py_uses_test_runner: no
     sphinx_docs: yes
     sphinx_docs_hosting: http://library-cookiecutters.gitlab.io/python-nameless
@@ -46,7 +46,7 @@ default_context:
     test_matrix_configurator: no
     test_matrix_separate_coverage: no
     test_runner: pytest
-    travis: no
+    travis: yes
```

**Comment by ionelmc** [comment]
Damn it, something is horribly wrong there - looks like it's not reading the previous settings (.cookiecutterrc) at all.

**Comment by dHannasch** [comment]
Working locally on my box, I tried hacking https://github.com/cookiecutter/cookiecutter/blob/master/cookiecutter/generate.py#L107:
```
    assert extra_context
    assert not extra_context['codecov']
    assert obj['codecov']
    assert context['cookiecutter']['codecov']
    assert obj is context['cookiecutter']
    if extra_context:
        apply_overwrites_to_context(obj, extra_context)
    assert context['cookiecutter']['codecov'][0] == 'yes'
    raise Exception(context['cookiecutter']['codecov'])
```
The asserts all pass.

Ah, I think I've got it. Cookiecutter, for whatever reason, silently swallows when a value for a choice variable is invalid. I've made a pull request for that.

**Comment by dHannasch** [comment]
If you're curious what was up, it's complicated.

I made a pull request on cookiecutter to update the YAML parser, and naturally enough my first thought was that maybe the only reason I was seeing errors with cookiepatcher was because the branch of cookiecutter I was using was performing differently somehow, so I pip-installed the master branch of cookiecutter to be sure.

But evidently, somehow, some way, it was still using my branch of cookiecutter. That part, I still haven't figured out. The logs *look* like it successfully installed the master branch of cookiecutter.

In any case, even using my yaml branch, how did that result in the observed behavior? Even if something was wrong with the parsing, how could that result in cookiepatcher ignoring the input file entirely?

The answer is that cookiecutter, for whatever reason, silently ignores incorrect inputs. (I made a pull request about *that* once I realized it.)

(How was my branch parsing the YAML wrong? Well, in a sense it wasn't. Some files can be parsed as YAML 1.1 or YAML 1.2, but don't parse the same way. cookiepatcher *writes* the .cookiecutterrc as YAML 1.2, but I was *reading* it as YAML 1.1. The reason this never came up before was that it just so happened that although poyo is this...not-quite-YAML-1.1-not-quite-YAML-1.2...thing...it was always parsing the same way as YAML 1.2 for the files you were parsing. When I updated cookiecutter to use a proper YAML parser, I picked a YAML 1.1 parser without thinking twice about it, and it turned out to be incompatible with a YAML 1.2 writer. Everything works fine with a YAML 1.2 parser.)
